Added region as an option for file-syncer

riltsken opened the pull request:
Before this it would default to the first region given from the
libcloud service catalog. If we pass a region to the driver then
it will use that region

Commit:
Added region as an option for file-syncer

Before this it would default to the first region given from the
libcloud service catalog. If we pass a region to the driver then
it will use that region

Kami commented on file_syncer/syncer.py:
```diff
@@ -99,7 +100,10 @@ def _setup_container(self):
-        driver = self._provider_cls(self._username, self._api_key)
+        driver = self._provider_cls(
+            self._username, self._api_key,
+            ex_force_service_region=self._region
```
This won't work if you use `--region` with a non-Rackspace / Swift provider, so I would do something like this:

``` python
PROVIDER_TO_REGION_KWARG_MAP = {
    'CLOUDFILES_US': 'ex_force_service_region',
    'CLOUDFILES_UK': 'ex_force_service_region'
}

args = (self._username, self._api_key)
kwargs = {}

provider_region_kwarg = PROVIDER_TO_REGION_KWARG_MAP.get(self._provider_name, None)
if self._region and region_kwarg:
    kwargs[provider_region_kwarg] = self._region

driver = self._provider_cls(*args, **kwargs)
```

Currently we only pass `provider_cls` to the constructor which means you will need to modify the code to also pass `provider_name` to it if you decide to go with this approach.

Kami commented on file_syncer/run.py:
```diff
@@ -39,6 +39,8 @@ def run():
     parser = OptionParser(usage=usage)
     parser.add_option('--provider', dest='provider', default='CLOUDFILES_US',
                       help='Provider to use')
+    parser.add_option('--region', dest='region', default=None,
```
I would change help text to say something like: `'Region to use if a Libcloud driver supports multiple regions (e.g. ORD for CloudFiles provider)`.

Commit:
Graceful fallback for non-rackspace / swift providers

Only try to add the region if the provider allows for it from
the given PROVIDER_TO_REGION mapping

riltsken commented:
Let me know if this addresses your concerns @Kami :)

Kami commented on file_syncer/syncer.py:
```diff
@@ -99,7 +102,22 @@ def _setup_container(self):
+        args = (self._username, self._api_key)
+        kwargs = {}
+
+        force_region = PROVIDER_HAS_REGION.get(self._provider)
```
This won't work because `options.provider` is a Libcloud provider type constant which is all uppercase. You can just do `PROVIDER_HAS_REGION.get(self._provider.lower(), None)`

riltsken replied:
Seemed to be working for me. Are they somehow converted from here? https://github.com/apache/libcloud/blob/trunk/libcloud/storage/types.py#L63

riltsken replied:
also do you prefer the explicit None versus implicit None?

Kami replied:
Hm, are you sure? Since the code uses keys and not values and the keys are all uppercase.

``` python
from libcloud.storage.types import Provider

SUPPORTED_PROVIDERS = [p for p in Provider.__dict__.keys() if not
                       p.startswith('__')]
```

You might also ask why do we use keys instead of values - in the older versions of Libcloud, values were plain numbers (e.g. `CLOUDFILES_US = 0`) so using keys was more user-friendly.

As far as `None` goes, yeah I prefer explicit `None`, but that's not a blocker or more of a personal preference :)

riltsken replied:
So it's actually grabbed from the PROVIDER_MAP and not the SUPPORTED_PROVIDERS. It just checks if that all caps version is in SUPPORTED_PROVIDERS. PROVIDER_MAP has keys as uppercase and values as lowercase.

https://gist.github.com/riltsken/4cd37630cae696fb8c11

Kami replied:
Ah, sorry, my bad then.

I thought `options.provider` is directly passed to the constructor, but it's not. Thanks for helping clear this up :)

Commit:
Explicit None when grabbing the forced region

riltsken commented:
Added the explicit None. Let me know what else I can do to get this in.

riltsken commented:
Haha nice :)

Kami commented:
Thanks, merged into master and `0.4.0` published to PyPi - https://pypi.python.org/pypi/file_syncer/0.4.0
